### DISCUSSION TITLE
How to leverage rdkit/Data/Functional_Group_Hierarchy.txt to match the functional groups in the molecules?

### DISCUSSION THREAD

**Discussion by DaShenZi721** [rdkit/rdkit#4605] [Q&A] [answered]
Hi All,

I am working on identifying functional groups of molecules with the functional groups' information from this file 'rdkit/Data/Functional_Group_Hierarchy.txt'.

Taking line 56 `  Halogen.Aromatic		[F,Cl,Br,I;$(*-!@c)]	Aromatic` for example, I used the following code to do matching

```
    mol = Chem.MolFromSmiles('[C:1]([CH:5]1[CH2:6][CH2:7][CH:8]([OH:11])[CH2:9][CH2:10]1)([CH3:4])([CH3:2])[CH3:3]')
    functional_group = Chem.MolFromSmarts('[F,Cl,Br,I;$(*-!@c)]')
    if functional_group is not None:
        match = mol.GetSubstructMatches(functional_group)
```

But it raises error

```
SMARTS Parse Error: syntax error while parsing: [F
SMARTS Parse Error: Failed parsing SMARTS '[F' for input: '[F'
```
Obviously, these SMARTS patterns are being used in the wrong way.

**Comment by dehaenw** [comment, marked as answer]
It may be a problem on your end. When I used the following code:
```python
mol = Chem.MolFromSmiles('c1ccccc1Br')
functional_group = Chem.MolFromSmarts('[F,Cl,Br,I;$(*-!@c)]')
match = mol.GetSubstructMatches(functional_group)
if len(match) > 0:
    print("there is a halogen attached to an aromatic atom in the molecule")
else:
    print("no match")
```
it executes without problem, giving the expected output, a match for bromobenzene. (and no match for your molecule 4tBuhexanol aka "patchouli hexanol")

Looking at your error trace you are probably opening Functional_Group_Hierarchy.txt using some csv reader and you have the delimiter set to ",". This is why during iteration it tries to put a SMARTS patterns "[F" which has an unclosed bracket isntead of taking the whole '[F,Cl,Br,I;$(*-!@c)]') SMARTS pattern that doesnt throw any error.

**Reply by DaShenZi721** [reply]
Hi @dehaenw!

Your analysis is so accurate. The problem lay with the csv reader. After calling 'readlines()' instead of csv reader, it executes without problem.

Thank you so much!

**Comment by bp-kelley** [comment]
The functional groups are also available as follows:

`>>> from rdkit.Chem import MolFromSmiles
>>> from rdkit.Chem.FilterCatalog import *
>>> functionalGroups = GetFunctionalGroupHierarchy()
>>> [match.filterMatch.GetName() 
...     for match in functionalGroups.GetFilterMatches(
...         MolFromSmiles('c1ccccc1Cl'))]
['Halogen.Aromatic', 'Halogen.NotFluorine.Aromatic'] `

to get the patterns themselves in a name:pattern_mol you can use

`rdkit.Chem.rdfiltercatalog.GetFlattenedFunctionalGroupHierarchy() `

**Reply by DaShenZi721** [reply]
Thank you for answering!

It will be helpful for my future work.
